gremlin-go/driver/connection.go

```diff
@@ -19,50 +19,62 @@ under the License.
+	established      connectionState = iota
+	closed           connectionState = iota
+	closedDueToError connectionState = iota
+)
+
 type connection struct {
-	host            string
-	port            int
-	transporterType TransporterType
-	logHandler      *logHandler
-	transporter     transporter
-	protocol        protocol
-	results         map[string]ResultSet
+	logHandler *logHandler
+	protocol   protocol
+	results    map[string]ResultSet
+	state      connectionState
 }
 
-func (connection *connection) close() (err error) {
-	if connection.transporter != nil {
-		err = connection.transporter.Close()
-	}
-	return
+func (connection *connection) errorCallback() {
+	connection.state = closedDueToError
+	_ = connection.protocol.close()
 }
 
-func (connection *connection) connect() error {
-	if connection.transporter != nil {
-		closeErr := connection.transporter.Close()
-		connection.logHandler.logf(Warning, transportCloseFailed, closeErr)
+func (connection *connection) close() (err error) {
+	if connection.state != established {
+		return errors.New("cannot close connection that has already been closed or has not been connected")
 	}
-	connection.protocol = newGremlinServerWSProtocol(connection.logHandler)
-	connection.transporter = getTransportLayer(connection.transporterType, connection.host, connection.port)
-	err := connection.transporter.Connect()
-	if err != nil {
-		return err
+	connection.logHandler.log(Info, closeConnection)
+	if connection.protocol != nil {
+		err = connection.protocol.close()
 	}
-	connection.protocol.connectionMade(connection.transporter)
-	return nil
+	connection.state = closed
+	return
 }
 
 func (connection *connection) write(request *request) (ResultSet, error) {
-	if connection.transporter == nil || connection.transporter.IsClosed() {
-		err := connection.connect()
-		if err != nil {
-			return nil, err
-		}
-	}
-	if connection.results == nil {
-		connection.results = map[string]ResultSet{}
-	}
+	connection.logHandler.log(Info, writeRequest)
+	requestID := request.requestID.String()
+	connection.logHandler.logger.Log(Info, "Creating request with id ", requestID)
+	connection.results[requestID] = newChannelResultSet(requestID)
+	return connection.results[requestID], connection.protocol.write(request)
+}
 
-	// Write through protocol layer.
-	responseID, err := connection.protocol.write(request, connection.results)
-	return connection.results[responseID], err
+func createConnection(logHandler *logHandler, host string, port int) (conn *connection, err error) {
+	conn = &connection{newLogHandler(&defaultLogger{}, Info, language.English), nil, map[string]ResultSet{}, initialized}
+	logHandler.log(Info, connectConnection)
+	conn.protocol, err = newGremlinServerWSProtocol(logHandler, Gorilla, host, port, conn.results, conn.errorCallback)
+	if err == nil {
+		conn.state = established
+		return conn, err
+	} else {
+		logHandler.logf(Error, failedConnection)
+		conn.state = closedDueToError
+		return nil, err
```

Comment:
This indicates that this is not a good time to use named return values.

In general, let's avoid them altogether. Their main benefit is for public-facing methods and can cause issues such as shadowing.